contracts/rust/src/bin/diff_test.rs

```diff
@@ -0,0 +1,1072 @@
+    fn from(c: ParsedChallenges) -> Self {
+        Self {
+            tau: Fr::from(0u32),
+            alpha: u256_to_field(c.alpha),
+            beta: u256_to_field(c.beta),
+            gamma: u256_to_field(c.gamma),
+            zeta: u256_to_field(c.zeta),
+            v: u256_to_field(c.v),
+            u: u256_to_field(c.u),
+        }
+    }
+}
+
+// modify from <https://github.com/EspressoSystems/cape/blob/main/contracts/rust/src/plonk_verifier/helpers.rs>
+/// return list of (proof, ver_key, public_input, extra_msg, domain_size)
+#[allow(clippy::type_complexity)]
+fn gen_plonk_proof_for_test(
+    num_proof: usize,
+) -> Result<
+    Vec<(
+        Proof<Bn254>,
+        VerifyingKey<Bn254>,
+        Vec<Fr>,
+        Option<Vec<u8>>,
+        usize,
+    )>,
+> {
+    // 1. Simulate universal setup
+    let rng = &mut jf_utils::test_rng();
+    let srs = {
+        let aztec_srs = crs::aztec20::kzg10_setup(1024)?;
+
+        UnivariateUniversalParams {
+            powers_of_g: aztec_srs.powers_of_g,
+            h: aztec_srs.h,
+            beta_h: aztec_srs.beta_h,
+        }
+    };
+    let open_key = open_key();
+    assert_eq!(srs.h, open_key.h);
+    assert_eq!(srs.beta_h, open_key.beta_h);
+    assert_eq!(srs.powers_of_g[0], open_key.g);
+
+    // 2. Create circuits
+    let circuits = (0..num_proof)
+        .map(|i| {
+            let m = 2 + i / 3;
+            let a0 = 1 + i % 3;
+            gen_circuit_for_test::<Fr>(m, a0)
+        })
+        .collect::<Result<Vec<_>>>()?;
+    let domain_sizes: Vec<usize> = circuits
+        .iter()
+        .map(|c| c.eval_domain_size().unwrap())
+        .collect();
+
+    // 3. Preprocessing
+    let mut prove_keys = vec![];
+    let mut ver_keys = vec![];
+    for c in circuits.iter() {
+        let (pk, vk) = PlonkKzgSnark::<Bn254>::preprocess(&srs, c)?;
+        prove_keys.push(pk);
+        ver_keys.push(vk);
+    }
+
+    // 4. Proving
+    let mut proofs = vec![];
+    let mut extra_msgs = vec![];
+
+    circuits
+        .iter()
+        .zip(prove_keys.iter())
+        .enumerate()
+        .for_each(|(i, (cs, pk))| {
+            let extra_msg = Some(format!("extra message: {}", i).into_bytes());
+            proofs.push(
+                PlonkKzgSnark::<Bn254>::prove::<_, _, SolidityTranscript>(
+                    rng,
+                    cs,
+                    pk,
+                    extra_msg.clone(),
+                )
+                .unwrap(),
+            );
+            extra_msgs.push(extra_msg);
+        });
+
+    let public_inputs: Vec<Vec<Fr>> = circuits
+        .iter()
+        .map(|cs| cs.public_input().unwrap())
+        .collect();
+
+    Ok(izip!(proofs, ver_keys, public_inputs, extra_msgs, domain_sizes).collect())
+}
+
+// Different `m`s lead to different circuits.
+// Different `a0`s lead to different witness values.
+fn gen_circuit_for_test<F: PrimeField>(m: usize, a0: usize) -> Result<PlonkCircuit<F>> {
```

Comment:
If this is just for testing purposes, maybe you can just panic here instead of returning a result?

Reply:
it is for testing only, I just find the syntax sugar `?` much easier to read than bunch of `unwrap()` when building the circuit. Consider:

```rust
    let c = cs.create_public_variable(
        (cs.witness(b[1])? + cs.witness(a[0])?) * (cs.witness(b[1])? - cs.witness(a[0])?),
    )?;

// v.s. 
    let c = cs.create_public_variable(
        (cs.witness(b[1]).unwrap() + cs.witness(a[0]).unwrap()) * (cs.witness(b[1]).unwrap() - cs.witness(a[0]).unwrap()),
    )?;
```

that said, I will turn more `?` into `expect("reason")` in the `fn gen_plonk_proof_for_test()` which is the wrapper for this `fn gen_circuit_for test` for faster panic with debuggable reasons.

Reply:
makes sense! I see what you're saying